### DISCUSSION TITLE
+append appears to reverse image order

### DISCUSSION THREAD

**Discussion by RoryLila** [ImageMagick/ImageMagick#6064] [Help]
### ImageMagick version

6.8.7.4

### Operating system, version and so on

Win XP

### Description

- A border is spliced to a set of rectangular images and the images are appended vertically. 
- The appended set is then cut into two vertical images.
- The two vertical images are then appended horizontally.
- However, after a normal append, the horizontal order of the two vertical segments is reversed.
- Thus, the normal order of the images for the horizontal append requires reversal, as below?

```
convert \( rectangles*.png -background black -splice 0x10 -append \
               -gravity south -background black -splice 0x10 \
               -write mpr:abc +delete \) \
             \( mpr:abc -crop 1141X2062+0+2052 \) \
             \( mpr:abc -crop 1141x2062+0+0    \) \
                +append ABC.png
```

**Comment by snibgo** [comment]
After every `+append` and `-append` and `-crop`, insert `+repage`. If any of your input images might have a canvas offset, also insert `+repage` after reading `rectangles*.png`.

Does that fix the problem?

**Comment by fmw42** [comment]
Please do not yell!

**Comment by RoryLila** [comment]
Thanks for your comment.

Upon the belief that for this script, `+repage' would "do no harm", I had
tried `+repage' on every line, to no avail. I then tried your suggestion
to insert `+repage' immediately after `rectangles*.png'. That did not
solve the problem either.

I have a feeling that `mpr:' is the culprit.

**Comment by RoryLila** [comment]
I am sorry if there was a perception of `yelling', however upon re-reading
I did not find anything I would think was a `yell'. I am interested to
know, so I can improve my communication style.

**Comment by fmw42** [comment]
Large bold font is considered to be yelling when used in excess.

**Comment by snibgo** [comment]
@RoryLila: to help us solve the problem, please show inputs and the output that shows the problem.

Your v6.8.7.4 is very old. I don't have a version that old. But I can try your command on a more recent version.

**Comment by RoryLila** [comment]
Oh! Sorry about that.

I am fairly deaf (and am told to stop shouting) and eyesight not that good
either, so even with glasses I often find 10 point font difficult to read
so I tend to increase it where I can, especially on my laptop.

**Comment by RoryLila** [comment]
Sorry Fred. I did not realise the font was so LARGE. Being a newbee to the
system, at time of entry it was either not large or I was nervous about
how to change it without risking losing the ability to enter my problem. I
had already taken much effort to just find a way to report Imagemagick 6
problems.

**Comment by fmw42** [comment]
There are two forums. One for IM 6 and one for IM 7. You posted an IM 6 issue in the IM 7 forum.  See https://github.com/ImageMagick/ImageMagick6 for IM 6 issues and https://github.com/ImageMagick/ImageMagick for IM 7 issues.  

Not to worry. We answer all messages even if in the wrong forum.

**Comment by RoryLila** [comment]
I sent this from another email site, however it may have not gone through.
So here it is again:

I have tried to strip the script to the minimal possible without +repage
options as they seemed to be not helping.

re.png, or.png, ye.png are each 200x200 images filled with red, orange,
yellow respectively.

First a test to see the vertical append, then a crop and horizontal append

convert \( re.png or.png ye.png -background black -splice 0x10 \) \
           -append -gravity south -background black -splice 0x10 ABC.png

convert \( \( re.png or.png ye.png -background black -splice 0x10 \) \
              -append -gravity south -background black -splice 0x10 \) \
              -write mpr:abc +delete \
           \( mpr:abc -crop 200X210+0+0   \) \
           \( mpr:abc -crop 200x210+0+210 \) \
           \( mpr:abc -crop 200x210+0+420 \)  +append DEF.png

The horizontal result is not in the expected order.


> @RoryLila: to help us solve the problem, please show inputs and the output
> that shows the problem.
>
> Your v6.8.7.4 is very old. I don't have a version that old. But I can try
> your command on a more recent version.
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/ImageMagick/ImageMagick/discussions/6064#discussioncomment-4989912
> You are receiving this because you were mentioned.
>
> Message ID:
> ***@***.***>


Damien
CMAD/Discway
634 Bell St
Preston Vic 3072
***@***.***

**Reply by fmw42** [reply]
Post your images so that we can test with them.

**Comment by snibgo** [comment]
ABC.png is as expected, with red at the top. Agreed?

DEF.png has yellow at the left. Perhaps you expect red to be at the left.

Perhaps you don't realise that "`-gravity south`" affects the behaviour of "`-crop`". It will crop from the *bottom* of the image instead of the top. Insert "`+gravity`" sometime before you crop anything.

**Comment by RoryLila** [comment]
Thank you Snibgo!

Imagemagick never fails to trick me! Yes, I expected red to be on the left.
I thought the outside \(...\) would isolate options in the top part from
the bottom and just pass on the image. Adding +gravity after +delete sets
the order with red on the left.

Interesting to learn that a list of images can be reversed by applying
-gravity in front.

**Comment by snibgo** [comment]
> Interesting to learn that a list of images can be reversed by applying -gravity in front.

"-gravity" doesn't reverse the list of images. Your first crop is:
```
 \( mpr:abc -crop 200X210+0+0   \) \
```
The default gravity would crop from the top of the image, so that would give you the red square. But with "-gravity south" the cropping is from the bottom, so that gives the yellow square.

This is useful for example:
```
magick in.png -gravity SouthEast -crop 200x100+0+0 +repage out.png
```
... to get the bottom-right of the input image. But we need to remember that "-gravity" affects all following operations. "+gravity" restores the default gravity.

**Comment by RoryLila** [comment]
Thanks for the correction. However, I made up a checker board for in.png
and using your command line with different coordinates for the -crop, I
could not correlate the appearance of out.png with -gravity positions, eg
South. The latter I assumed to be the midpoint of the bottom edge.
